src/udev/memory_id/dmidecode.c

```diff
@@ -0,0 +1,768 @@
+ * information does not come from the above mentioned specification.
+ *
+ * Additional references:
+ *  - Intel AP-485 revision 36
+ *    "Intel Processor Identification and the CPUID Instruction"
+ *    http://www.intel.com/support/processors/sb/cs-009861.htm
+ *  - DMTF Common Information Model
+ *    CIM Schema version 2.19.1
+ *    http://www.dmtf.org/standards/cim/
+ *  - IPMI 2.0 revision 1.0
+ *    "Intelligent Platform Management Interface Specification"
+ *    http://developer.intel.com/design/servers/ipmi/spec.htm
+ *  - AMD publication #25481 revision 2.28
+ *    "CPUID Specification"
+ *    http://www.amd.com/us-en/assets/content_type/white_papers_and_tech_docs/25481.pdf
+ *  - BIOS Integrity Services Application Programming Interface version 1.0
+ *    http://www.intel.com/design/archives/wfm/downloads/bisspec.htm
+ *  - DMTF DSP0239 version 1.1.0
+ *    "Management Component Transport Protocol (MCTP) IDs and Codes"
+ *    http://www.dmtf.org/standards/pmci
+ *  - "TPM Main, Part 2 TPM Structures"
+ *    Specification version 1.2, level 2, revision 116
+ *    https://trustedcomputinggroup.org/tpm-main-specification/
+ *  - "PC Client Platform TPM Profile (PTP) Specification"
+ *    Family "2.0", Level 00, Revision 00.43, January 26, 2015
+ *    https://trustedcomputinggroup.org/pc-client-platform-tpm-profile-ptp-specification/
+ *  - "RedFish Host Interface Specification" (DMTF DSP0270)
+ *    https://www.dmtf.org/sites/default/files/DSP0270_1.0.1.pdf
+ */
+
+#include <stdio.h>
+#include <string.h>
+#include <stdlib.h>
+#include <getopt.h>
+#define _GNU_SOURCE
+#include <errno.h>
+
+#include "build.h"
+#include "fileio.h"
+#include "udev-util.h"
+#include "types.h"
+
+#define out_of_spec "<OUT OF SPEC>"
+static const char *bad_index = "<BAD INDEX>";
+
+#define SUPPORTED_SMBIOS_VER 0x030300
+
+#define FLAG_NO_FILE_OFFSET     (1 << 0)
+#define FLAG_STOP_AT_EOT        (1 << 1)
+#define FLAG_FROM_DUMP          (1 << 2)
+static uint32_t opt_flags = 0;
+
+#define SYS_FIRMWARE_DIR "/sys/firmware/dmi/tables"
+#define SYS_ENTRY_FILE SYS_FIRMWARE_DIR "/smbios_entry_point"
+#define SYS_TABLE_FILE SYS_FIRMWARE_DIR "/DMI"
+
+static int checksum(const uint8_t *buf, size_t len) {
+        uint8_t sum = 0;
+        size_t a;
+
+        for (a = 0; a < len; a++)
+                sum += buf[a];
+        return (sum == 0);
+}
+
+/*
+ * Type-independant Stuff
+ */
+
+static char *_dmi_string(const struct dmi_header *dm, uint8_t s)
+{
+        char *bp = (char *)dm->data;
+
+        bp += dm->length;
+        while (s > 1 && *bp)
+        {
+                bp += strlen(bp);
+                bp++;
+                s--;
+        }
+
+        if (!*bp)
+                return NULL;
+
+        return bp;
+}
+
+static const char *dmi_string(const struct dmi_header *dm, uint8_t s)
+{
+        char *bp;
+
+        if (s == 0)
+                return "Not Specified";
+
+        bp = _dmi_string(dm, s);
+        if (bp == NULL)
```

Comment:
doesn't matter, but for ptr != NULL checks we usually rely on C's downgrade-to-bool feature, i.e. we'd write this as `if (!bp)…`

we insist on explicit comparisons only for variables that actually arithmetic inetegres or floats, i.e. not pointers, bools or boolean-like.